migrations/sqls/20191101101124-billing-tables-up.sql

```diff
@@ -0,0 +1,156 @@
+
+
+-- reinstate the foreign key relationships
+alter table if exists water.charge_agreements
+  add constraint charge_agreements_charge_element_id_fkey
+    foreign key (charge_element_id)
+    references water.charge_elements (charge_element_id)
+    match simple
+    on delete cascade;
+
+alter table if exists water.charge_elements
+  add constraint charge_elements_charge_version_id_fkey
+    foreign key (charge_version_id)
+    references water.charge_versions (charge_version_id)
+    match simple
+    on delete cascade;
+
+
+
+------------------------
+-- create the new tables
+------------------------
+
+create table water.billing_batches (
+  billing_batch_id uuid primary key default gen_random_uuid(),
+  region_id uuid not null
+    constraint billing_batches_region_id_fkey
+    references water.regions (region_id),
+  batch_type water.charge_batch_type not null,
+  financial_year integer not null,
+  season water.charge_element_season not null,
+  date_created timestamp not null default now(),
+  date_updated timestamp not null default now()
+);
+
+create table water.billing_batch_charge_versions (
+  billing_batch_charge_version_id uuid primary key default gen_random_uuid(),
+  billing_batch_id uuid not null
+    constraint billing_batch_charge_versions_billing_batch_id_fkey
+    references water.billing_batches (billing_batch_id),
+  charge_version_id uuid not null
+    constraint billing_batch_charge_versions_charge_version_id_fkey
+    references water.charge_versions (charge_version_id),
+  date_created timestamp not null default now(),
+  date_updated timestamp not null default now()
+);
+
+create table water.billing_invoices (
+  billing_invoice_id uuid primary key default gen_random_uuid(),
+  invoice_account_id uuid not null,
+  address jsonb not null,
+  invoice_account_number text,
+  net_amount decimal,
+  is_credit boolean,
+  date_created timestamp not null default now(),
+  date_updated timestamp not null default now()
+);
+
+create table water.billing_batch_invoices (
+  billing_batch_invoice_id uuid primary key default gen_random_uuid(),
+  billing_batch_id uuid not null
+    constraint billing_batch_invoices_billing_batch_id_fkey
+    references water.billing_batches (billing_batch_id),
+  billing_invoice_id uuid not null
+    constraint billing_batch_invoices_billing_invoice_id_fkey
+    references water.billing_invoices (billing_invoice_id),
+  date_created timestamp not null default now(),
+  date_updated timestamp not null default now()
+);
+
+create table water.billing_invoice_licences (
+  billing_invoice_licence_id uuid primary key default gen_random_uuid(),
+  billing_invoice_id uuid not null
+    constraint billing_invoice_licences_billing_invoice_id_fkey
+    references water.billing_invoices (billing_invoice_id),
+  company_id uuid not null,
+  contact_id uuid not null,
+  address_id uuid not null,
+  licence_ref varchar not null,
```

Comment:
should licence_ref also have a foreign key constraint?

Reply:
I didn't think so really, because licence_ref would map to a table outside of the current schema.

Reply:
So not required due to SOA.